File: cppLeetcode/leetcodeM/935_KnightDialer.cpp
/*
 935. Knight Dialer
 
 
 The chess knight has a unique movement, it may move two squares vertically and one square horizontally, or two squares horizontally and one square vertically (with both forming the shape of an L).
 
 The possible movements of chess knight are shown in this diagaram:

 A chess knight can move as indicated in the chess diagram below:


 We have a chess knight and a phone pad as shown below, the knight can only stand on a numeric cell (i.e. blue cell).


 Given an integer n, return how many distinct phone numbers of length n we can dial.

 You are allowed to place the knight on any numeric cell initially and then you should perform n - 1 jumps to dial a number of length n. All jumps should be valid knight jumps.

 As the answer may be very large, return the answer modulo 109 + 7.

  

 Example 1:

 Input: n = 1
 Output: 10
 Explanation: We need to dial a number of length 1, so placing the knight over any numeric cell of the 10 cells is sufficient.
 Example 2:

 Input: n = 2
 Output: 20
 Explanation: All the valid number we can dial are [04, 06, 16, 18, 27, 29, 34, 38, 40, 43, 49, 60, 61, 67, 72, 76, 81, 83, 92, 94]
 Example 3:

 Input: n = 3
 Output: 46
 Example 4:

 Input: n = 4
 Output: 104
 Example 5:

 Input: n = 3131
 Output: 136006598
 Explanation: Please take care of the mod.
  

 Constraints:

 1 <= n <= 5000

 
 */
#include <vector>
#include <algorithm>
#include <numeric>

using std::vector;

class knightDialer {
    
public:
    
    int doit_dp(int n) {
        
        vector<vector<int>> jump{{4, 6}, {6, 8}, {7, 9}, {4, 8}, {0, 3, 9}, {}, {0, 1, 7}, {2, 6}, {1, 3}, {2, 4}};
        int mod = 1e9+7;
        vector<int> dp(10, 1);
        
        while (n>1) {
                
            vector<int> nextdp(10, 0);
            
            for (auto i = 0; i < dp.size(); i++) {
                for (auto j : jump[i])
                    nextdp[j] = (nextdp[j] + dp[i]) % mod;
            }
            nextdp.swap(dp);
            n--;
        }
        
        int res = 0;
        for (auto n : dp)
            res = (res + n) % mod;
        
        return res;
    }
    
    int doit_dp_fastpow(int n) {
        
        vector<vector<long>> ans(1, vector<long>(10, 1));
        
        vector<vector<long>> matrix{
            {0, 0, 0, 0, 1, 0, 1, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 1, 0, 1, 0},
            {0, 0, 0, 0, 0, 0, 0, 1, 0, 1},
            {0, 0, 0, 0, 1, 0, 0, 0, 1, 0},
            {1, 0, 0, 1, 0, 0, 0, 0, 0, 1},
            {0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
            {1, 1, 0, 0, 0, 0, 0, 1, 0, 0},
            {0, 0, 1, 0, 0, 0, 1, 0, 0, 0},
            {0, 1, 0, 1, 0, 0, 0, 0, 0, 0},
            {0, 0, 1, 0, 1, 0, 0, 0, 0, 0}
        };
        int mod = 1e9 + 7;
        
        auto mul = [mod](const vector<vector<long>>& A, const vector<vector<long>>& B) {
            
            const auto m = A.size(), n = B.size();
            const auto p = B[0].size();
            
            vector<vector<long>> C(m, vector<long>(p));
            for (auto i = 0; i < m; i++) {
                for (auto j = 0; j < p; j++) {
                    for (auto k = 0; k < n; k++) {
                        C[i][j] = (C[i][j] + A[i][k] * B[k][j]) % mod;
                    }
                }
            }
            return C;
        };
        
        n--;
        while (n) {
            if (n & 1)
                ans = mul(ans, matrix);
            
            matrix = mul(matrix, matrix);
            n >>= 1;
        }
        
        long res = 0;
        for (auto c : ans[0])
            res = (res + c) % mod;
        
        return res;
    }
};

File: PythonLeetcode/leetcodeM/935_KnightDialer.py
"""
935. Knight Dialer

# A chess knight can move as indicated in the chess diagram below:

# This time, we place our chess knight on any numbered key of a phone pad (indicated above), and the knight makes N-1 hops.
# Each hop must be from one key to another numbered key.

# Each time it lands on a key (including the initial placement of the knight), it presses the number of that key, pressing N digits total.

# How many distinct numbers can you dial in this manner?

# Since the answer may be large, output the answer modulo 10^9 + 7.


# Example 1:

# Input: 1
# Output: 10

# Example 2:

# Input: 2
# Output: 20

# Example 3:

# Input: 3
# Output: 46


# Note:

# 1 <= N <= 5000
"""


class KnightDialer:

    """
    Approach 1: Dynamic Programming
    Intuition

    Let f(start, n) be the number of ways to dial an n digit number, where the knight starts at square start.
    We can create a recursion, writing this in terms of f(x, n-1)'s.

    Algorithm

    By hand or otherwise, have a way to query what moves are available at each square.
    This implies the exact recursion for f. For example, from 1 we can move to 6, 8, so f(1, n) = f(6, n-1) + f(8, n-1).

    After, let's keep track of dp[start] = f(start, n), and update it for each n from 1, 2, ..., N.

    At the end, the answer is f(0, N) + f(1, N) + ... + f(9, N) = sum(dp).

    Complexity Analysis

    Time Complexity: O(N).

    Space Complexity: O(1).

    """

    def doit_dp(self, N):

        keys_hop = [[4, 6], [6, 8], [7, 9], [4, 8], [0, 3, 9], [], [0, 1, 7], [2, 6], [1, 3], [2, 4]]
        MOD = 10 ** 9 + 7
        dp = [1] * 10

        for _ in range(N-1):
            dp2 = [0] * 10
            for n, cnt in enumerate(dp):
                for c in keys_hop[n]:
                    dp2[c] += cnt
                    dp2[c] %= MOD
            dp = dp2
        return sum(dp) % MOD

    def doit_dp_1(self, N):

        keys = {0: [4, 6], 1: [6, 8], 2: [7, 9], 3: [4, 8], 4: [0, 3, 9], 5: [], 6: [0, 1, 7], 7: [2, 6], 8: [1, 3], 9: [2, 4]}
        MOD = 10 ** 9 + 7
        memo = {}

        def hop(i, times):

            if (i, times) in memo:
                return memo[(i, times)]

            if times == 1:
                return [str(i)]

            ans = []
            for c in keys[i]:
                for w in hop(c, times-1):
                    ans.append(str(i) + w)
            memo[(i, times)] = ans
            return ans

        ans = []
        for i in range(10):
            ans.extend(hop(i, N))

        return len(ans) % MOD


if __name__ == '__main__':

    res = KnightDialer().doit(1)  # 10

    res = KnightDialer().doit(2)  # 20

    res = KnightDialer().doit(3)  # 46
